client/src/components/Users.js

```diff
@@ -413,9 +415,9 @@ function Users(props) {
         </TableCell>
         <TableCell>{user.active ? 'active' : 'inactive'}</TableCell>
         <TableCell>
-          {user.role.map((r, i) => (
-            <Grid key={i}>
-              {permissions.reduce((a, c) => a || (c.id === r ? c : undefined), undefined).roleName}
+          {user.role.map((r, idx) => (
+            <Grid key={`role_${idx}`}>
+              {permissions.reduce((a, c) => a || (c.id === r ? c.roleName : undefined), undefined)}
```

Comment:
I noticed that some users have role IDs 5 and 6, which no longer exist as permissions. Some have these roles multiple times. I considered turning permissions into a key/value map or array to make this process faster. map * reduce is O(m * n), but map * lookup is O(m). More of a comment, not a thing that needs done. It also would be nice to have the invalid ones filtered out rather than mapping over those values and generating JSX tags

Reply:
I agree, this certainly needs tidying up at some point.